File: DatasetGenerator.cpp
#include <iostream>
#include <fstream>
#include <unordered_set>
#include <random>
#include <string>
#include <vector>
#include <algorithm>
#include <sys/stat.h> // For mkdir on POSIX/Windows
#include <sys/types.h>
#include <io.h>

int main() {
    const int MAX_VALUE = 1000000000;
    const int DATASET_SIZE = 1000000;
    const int MIN_STRING_LENGTH = 4;
    const int MAX_STRING_LENGTH = 6;
    const std::string CHARACTERS = "abcdefghijklmnopqrstuvwxyz";
    const std::string DIR_NAME = "dataset";
    const std::string FILE_NAME = "dataset_cpp.csv";
    
    // Create dataset directory if it doesn't exist
    #if defined(_WIN32)
        _mkdir(DIR_NAME.c_str());
    #else
        mkdir(DIR_NAME.c_str(), 0777);
    #endif
    std::unordered_set<int> unique_numbers;
    std::vector<std::pair<int, std::string>> dataset;
    dataset.reserve(DATASET_SIZE);

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> num_dist(1, MAX_VALUE);
    std::uniform_int_distribution<> len_dist(MIN_STRING_LENGTH, MAX_STRING_LENGTH);
    std::uniform_int_distribution<> char_dist(0, CHARACTERS.size() - 1);

    while (dataset.size() < DATASET_SIZE) {
        int random_number = num_dist(gen);
        if (unique_numbers.find(random_number) != unique_numbers.end()) continue;
        unique_numbers.insert(random_number);

        int string_length = len_dist(gen);
        std::string random_string;
        random_string.reserve(string_length);
        for (int i = 0; i < string_length; ++i) {
            random_string += CHARACTERS[char_dist(gen)];
        }

        dataset.emplace_back(random_number, random_string);
    }

    // Shuffle dataset
    std::shuffle(dataset.begin(), dataset.end(), gen);

    // Write to CSV
    std::ofstream file(DIR_NAME + "/" + FILE_NAME);
    if (!file) {
        std::cerr << "Failed to open file for writing.\n";
        return 1;
    }
    for (const auto& pair : dataset) {
        file << pair.first << ',' << pair.second << '\n';
    }
    file.close();

    std::cout << "Dataset generated and written to " << DIR_NAME << "/" << FILE_NAME << std::endl;
    return 0;
}

File: DatasetGenerator.py
import os
import random
import csv

def generate_dataset(size, max_value, min_string_length, max_string_length):
    dataset = []
    characters = "abcdefghijklmnopqrstuvwxyz"
    unique_numbers = set()

    while len(dataset) < size:
        random_number = random.randint(1, max_value)  # Ensure positive numbers

        # Generate random string
        string_length = random.randint(min_string_length, max_string_length)
        random_string = ''.join(random.choice(characters) for _ in range(string_length))

        # Check if number is unique
        if random_number not in unique_numbers:
            unique_numbers.add(random_number)
            dataset.append((random_number, random_string))

    # Shuffle the dataset to ensure random order
    random.shuffle(dataset)
    return dataset

def write_dataset_to_file(dataset, file_name):
    # Create dataset directory if it doesn't exist
    os.makedirs("dataset", exist_ok=True)

    with open(f"dataset/{file_name}", mode="w", newline="") as file:
        writer = csv.writer(file)
        writer.writerows(dataset)

def main():
    # Define the maximum value for the integers
    MAX_VALUE = 1_000_000_000
    # Define the size of the dataset
    DATASET_SIZE = 1_000_000  # Testing size
    # Define string length range
    MIN_STRING_LENGTH = 4
    MAX_STRING_LENGTH = 6

    # Generate the dataset
    dataset = generate_dataset(DATASET_SIZE, MAX_VALUE, MIN_STRING_LENGTH, MAX_STRING_LENGTH)

    # Write the dataset to a file
    output_file_name = "dataset_py.csv"
    local_file_path = f"dataset/{output_file_name}"
    try:
        write_dataset_to_file(dataset, output_file_name)
        print(f"Dataset generated and written to {local_file_path}")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()